### DISCUSSION TITLE
Use Morphologist from Python instead of using GUI

### DISCUSSION THREAD

**Discussion by pierpy** [brainvisa/brainvisa.github.io#52] [Q&A] [answered]
Is it possible to run the morphologist pipeline as a python script?
If yes, where can I find some documentation about ?

**Comment by denisri** [comment, marked as answer]
Hi,
Yes there is the `axon-runprocess` command, see https://brainvisa.info/axon/user_doc/axon_manual2.html#distributed-execution-with-batch-mode, and there are also ways to import libraries and start a process from a python program:
```
from brainvisa.axon.processes import initializeProcesses
initializeProcesses()
from brainvisa.processes import defaultContext
defaultContext.runProcess('threshold', 'in_image.nii', 'out_image.nii.gz', mode='lt', threshold1=100)
```
You can get the doc of individual processes either in the brainvisa GUI, or using:
```
axon-runprocess --process-help threshold
```

**Reply by f4bry** [reply]
Nice! Is it possible to set up the python interpreter (brainvisa environment) when using singularity to install brainvisa?

**Comment by denisri** [comment]
When using the singularity container, the python environment is already setup: the brainvisa modules are already in the python path. Your python scripts just need to import the modules as explained above.

**Reply by f4bry** [reply]
Got it! It works if I run `bv bash`. 
FYI, I am on a Mac OS X 11.5, running singularity with the vagrant box. 
Thanks!

**Reply by sarafv** [reply]
Hi,
It is better to use axon instead of brainvisa process?
Here my script to run morphologist :

      brainvisaScriptFile = os.path.join(tempDir,'temp.py') 

      script = open(brainvisaScriptFile,'w')
      script.write("from brainvisa.processes import *\n")
      script.write("context = defaultContext()\n")
      script.write("context.runProcess('Morphologist',t1mri='%s',perform_normalization=False)\n" % self.mriPath)
      script.close 

then the script is launch in this way :  

>>  brainvisa -b -e  /tmp/tmpim02w98d/temp.py

**Reply by denisri** [reply]
It's the same. Or I don't get your question ? The only difference is that in a python script you have to import brainvisa.axon libs and initialize the engine (as in the example script I gave you above), and in your last message you are using "brainvisa -e" which runs the script within a brainvisa environment already initialized.

**Reply by sarafv** [reply]
Thank you!
So if it is not more efficient to do the same with axon I'll keep my way?

Sara

**Comment by denisri** [comment]
As you like...

**Reply by Sofiia2001** [reply]
@denisri Hello! I just came across your question and I was wondering if you managed to run your processes from python env on singularity? I am using the bv bash python environment and I cannot import anything on python 3 because it gives me weird errors of some "missing modules". When I import the processes module in python 2, it does not give an error, but throws one when I try to import the ParametrizeSulcus2015.py process, it says something like

```
>>> processes.readProcess('ParameterizeSulcus2015.py')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/casa/install/python/brainvisa/processes.py", line 6168, in readProcess
    if _readProcessLog is not None:
NameError: global name '_readProcessLog' is not defined
```

Did you manage to setup python so that everything imports and works OK?

Thank you in advance!
